smartsim/_core/entrypoints/dragon.py

```diff
@@ -0,0 +1,178 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import argparse
+import json
+import os
+import signal
+import socket
+import textwrap
+import typing as t
+from types import FrameType
+
+import zmq
+
+import smartsim._core.utils.helpers as _helpers
+from smartsim._core.launcher.dragon.dragonBackend import DragonBackend
+from smartsim._core.schemas import (
+    DragonBootstrapRequest,
+    DragonBootstrapResponse,
+    DragonShutdownResponse,
+)
+from smartsim._core.schemas.dragonRequests import request_serializer
+from smartsim._core.schemas.dragonResponses import response_serializer
+from smartsim._core.utils.network import get_best_interface_and_address
+
+# kill is not catchable
+SIGNALS = [signal.SIGINT, signal.SIGQUIT, signal.SIGTERM, signal.SIGABRT]
+
+SHUTDOWN_INITIATED = False
+
+
+def handle_signal(signo: int, _frame: t.Optional[FrameType]) -> None:
+    if not signo:
+        print("Received signal with no signo")
+    else:
+        print(f"Received {signo}")
+    cleanup()
+
+
+context = zmq.Context()
+
+"""
+Dragon server entrypoint script
+"""
+
+DBPID: t.Optional[int] = None
+
+
+def print_summary(network_interface: str, ip_address: str) -> None:
+    zmq_config = {"interface": network_interface, "address": ip_address}
+
+    with open("dragon_config.log", "w", encoding="utf-8") as dragon_config_log:
+        dragon_config_log.write(
+            textwrap.dedent(f"""\
+                -------- Dragon Configuration --------
+                IPADDRESS: {ip_address}
+                NETWORK: {network_interface}
+                HOSTNAME: {socket.gethostname()}
+                DRAGON_SERVER_CONFIG: {json.dumps(zmq_config)}
+                --------------------------------------
+
+                --------------- Output ---------------
+
+                """),
+        )
+
+
+def run(dragon_head_address: str) -> None:
+    global SHUTDOWN_INITIATED  # pylint: disable=global-statement
+    print(f"Opening socket {dragon_head_address}")
+    dragon_head_socket = context.socket(zmq.REP)
+    dragon_head_socket.bind(dragon_head_address)
+    dragon_backend = DragonBackend()
+
+    while not SHUTDOWN_INITIATED:
+        print(f"Listening to {dragon_head_address}")
+        req = dragon_head_socket.recv_json()
+        print(f"Received request: {req}")
+        drg_req = request_serializer.deserialize_from_json(str(req))
```

Comment:
I think the `SchemaSerializer` might not be the best choice. Pydantic should give us a ton of power over JSON serialization behaviors already.

What's wrong with built-ins?

```
import pydantic


class Book(pydantic.BaseModel):
    title: str
    num_pages: int


book = Book(title="The shortest title in the entire world", num_pages=1)

json = book.model_dump_json()
dict_ = book.model_dump()

print(json)
print(dict_)


```

output:

```bash
-- /Users/chris.mcbride/code/zmq_multi/.vscode/pydantic_serialization.py 
{"title":"The shortest title in the entire world","num_pages":1}
{'title': 'The shortest title in the entire world', 'num_pages': 1}
```

Reply:
The problem with the pydantic serialization is that it does not encode any type information about the original schemas, so if we want the dragon backend to be able to dispatch based on message type, we need some additional entity to provide it (it cannot be an instance attr on the pydantic schema itself, like it was originally, as pydantic will happly override that "defualt" meaning you could end up in a situation where you have a `DragonStartRequest` where `dragon_start_request.msg_type == "dragon_stop_request"`). Under the hood, all the schema serializer is doing is adding a `"__{request,response}_type__"` type key to the dict that pydantic natively dumps out.

The only other way I could think to encode this info would be to set up a req/res channel for each message type, which I could be swayed to, but I would be VERY hesitant to have pydantic encode type information give how passive and coercive it tends to be.

Reply:
then the schema is missing data it needs. 

we should add something like:

```
@computed_field
def message_type(self) -> str:
  return type(self).__name__
```

Reply:
Sure, but unless we want to devolve into `eval(payload["message_type"]).parse_obj(payload)`, we still need a mapping of "payload type strings" to "fns to parse the payload" and at that point I would argue that we just remade the `SchemaSerializer` class

Reply:
then make the subclasses implement it:

` 
import abc
import pydantic
import typing as t


class Base(pydantic.BaseModel):

    @pydantic.computed_field
    def message_type(self) -> str:
        return self.message_key()

    @staticmethod
    @abc.abstractmethod
    def message_key() -> str: ...


class Book(Base):
    title: str
    num_pages: int

    @staticmethod
    def message_key() -> str:
        return "book"


class Car(Base):
    make: str
    model: str

    @staticmethod
    def message_key() -> str:
        return "car"


book = Book(title="The shortest title in the entire world", num_pages=1)
car = Car(make="Ford", model="F150")

book_json = book.model_dump_json()
car_json = car.model_dump_json()


print(book_json)
print(car_json)


def deserialize_message(message: str) -> t.Optional[Base]:
    for msg_class in [Book, Car]:
        try:
            item = msg_class.model_validate_json(message)
            return item
        except pydantic.ValidationError:
            ...

    return None


obj1 = deserialize_message(book_json)
obj2 = deserialize_message(car_json)

print(obj1)
print(obj2)

`

Reply:
Even easier... If we need to filter messages, we can also just swap to a different zmq socket type. PUB/SUB pairs let us set up an explicit filter...

Or, set up a socket for each message type and be dirt simple?

Plenty of options without rebuilding pydantic deserialization .

Reply:
Per offline conversation:

WRT the first example: This works GREAT iff models have distinct sets of attrs w/o default values. However, if two models have the same shape, for e.g.

```py
import abc
import pydantic
import typing as t


class Base(pydantic.BaseModel):

    @pydantic.computed_field
    def message_type(self) -> str:
        return self.message_key()

    @staticmethod
    @abc.abstractmethod
    def message_key() -> str: ...


class Book(Base):
    title: str = "default"   # <-- NOTE: new default values here 
    num_pages: int = 123     #           and here!

    @staticmethod
    def message_key() -> str:
        return "book"


class Car(Base):
    make: str = "car brand"     # <-- NOTE: new default values here 
    model: str = "the new one"  #           and here!

    @staticmethod
    def message_key() -> str:
        return "car"


book = Book(title="The shortest title in the entire world", num_pages=1)
car = Car(make="Ford", model="F150")

book_json = book.model_dump_json()
car_json = car.model_dump_json()


print(book_json)
print(car_json)


def deserialize_message(message: str) -> t.Optional[Base]:
    for msg_class in [Book, Car]:
        try:
            item = msg_class.model_validate_json(message)
            return item
        except pydantic.ValidationError as e:
            print(e)

    return None


obj1 = deserialize_message(book_json)
obj2 = deserialize_message(car_json)

print(type(obj1))
print(obj1)
print('============')
print(type(obj2))
print(obj2)
```
will erroneously print out two books. And right now we have multiple schemas that violate that precondition, which runs into the same problem, which is why I am still VERY hesitant to allow pydantic handle de-serialization alone. I think it is too passive.

---

WRT to option two, I would argue that is functionally equivalent to
```py
str_to_cls_map = {cls.message_type: cls for cls in (Book, Car)}
#                     ^^^^^^^^^^^^ For sake of argument assume that this a class property

def load(payload):
    d = json.loads(message_json)
    return str_to_cls_map[d['message_type']].model_validate_json(d)
```
which I would argue is near identical to `SchemaSerializer.deserialize_from_json`.

----

Currently exploring setting up different ZMQ channels to route specific requests based on type and removing the need for the `SchemaSerializer` entirely. Failing that, trying to reduce the scope of `SchemaSerializer` to explicitly work on hydrating requests only (no serialization).

Reply:
I believe you guys reached a conclusion, for which a PR is up already, please feel free to resolve this conversation if it is correct.